### PR TITLE
Project update for compliance and client runs elasticsearch bug

### PR DESCRIPTION
**Pull request by lancewf**
### :nut_and_bolt: Description: What code changed, and why?
When running a project update with a chef tag project rule the compliance-service would fail to update. The problem is that the array field is sometimes stored as "null" and not an empty array. The solution is to check if the field is null before using it. 

Client runs had the same problem which is fixed in the PR also. 

### :chains: Related Resources

#1494

### :+1: Definition of Done
Able to successfully run a project update with a project with chef tag ingest rules when there are compliance reports with empty chef tags. 

### :athletic_shoe: How to Build and Test the Change
1. `build components/compliance-service && start_all_services`
1. Add some compliance nodes with `chef_load_compliance_nodes 100`
1. Open the compliance page https://a2-dev.test/compliance/reports/nodes
1. Copy one of the node's environment, its displayed in the list view
    - or use the environment filter at the top of the page to identify what a node's enviornment is
1. Create a project with two rules. One matching on a chef tag with any value. The second rule matching the environment found above. 
1. Update the projects
1. Go back to the https://a2-dev.test/compliance/reports/nodes page and filter by the project created to ensure the project update worked correctly. 

### :white_check_mark: Checklist

- [x] Tests added/updated?
- [ ] Docs added/updated?

### :camera: Screenshots, if applicable

### CODE REVIEW

File: components/compliance-service/ingest/ingestic/ingestic.go [hunk 1]
```diff
@@ -284,6 +284,10 @@ func (backend *ESClient) UpdateReportProjectsTags(ctx context.Context, projectTa
 						}
 					}
 					if (condition.roles.length > 0) {
+						if (ctx._source['roles'] == null) {
```
Comment (lancewf) [inline components/compliance-service/ingest/ingestic/ingestic.go]: The condition does not match if its type is roles and there are no roles on the report.

File: components/compliance-service/ingest/ingestic/ingestic.go [hunk 2]
```diff
@@ -300,6 +304,10 @@ func (backend *ESClient) UpdateReportProjectsTags(ctx context.Context, projectTa
 						}
 					}
 					if (condition.chefTags.length > 0) {
+						if (ctx._source['chef_tags'] == null) {
+							match = false;
+							break;
+						}
 						def found = false;
 						for (def ruleChefTag : condition.chefTags) {
 							for (def ccrChefTag : ctx._source.chef_tags) {
```
Comment (lancewf) [inline components/compliance-service/ingest/ingestic/ingestic.go]: This where the error was thrown. When the `ctx._source.chef_tags` == null

File: components/compliance-service/integration_test/reports_projects_update_test.go [hunk 3]
```diff
@@ -410,6 +410,30 @@ func TestProjectUpdate(t *testing.T) {
 			},
 			projectIDs: []string{},
 		},
+		{
```
Comment (lancewf) [inline components/compliance-service/integration_test/reports_projects_update_test.go]: These two tests failed with the same elasticsearch error from the issue before the above changes.

File: components/compliance-service/ingest/ingestic/ingestic.go [hunk 4]
```diff
@@ -416,6 +424,10 @@ func (backend *ESClient) UpdateSummaryProjectsTags(ctx context.Context, projectT
 						}
 					}
 					if (condition.roles.length > 0) {
```
Comment (msorens) [inline components/compliance-service/ingest/ingestic/ingestic.go]: Since each attribute has an identical code block pattern--and indeed identical code across `UpdateReportProjectsTags`, `UpdateSummaryProjectsTags`, and `UpdateNodeProjectTags`--you could reduce a lot of duplication by introducing two simple code generation functions, so that `script` in this function could be reduced to this, making it easier to maintain:
```
script := `
		ArrayList matchingProjects = new ArrayList();
		for (def project : params.projects) {
			for (def rule : project.rules) {
				def match = rule.conditions.length != 0;
				for (def condition : rule.conditions) {` +
					genSingleLoop("chefServer", "source_fqdn") + 
					genSingleLoop("environment", "environment") + 
					genSingleLoop("organization", "organization_name") + 
					genSingleLoop("policyName", "policy_name") + 
					genSingleLoop("policyGroup", "policy_group") + 
					genDoubleLoop("roles", "roles", "Role") + 
					genDoubleLoop("chefTags", "chef_tags", "ChefTag") + `
				}
				if ( match ) {
					matchingProjects.add(project.name);
					break;
				}
			}
		}
		ctx._source.projects = matchingProjects.toArray();
`
```

File: components/compliance-service/ingest/ingestic/ingestic.go [hunk 5]
```diff
@@ -500,6 +516,12 @@ func (backend *ESClient) JobStatus(ctx context.Context, jobID string) (project_u
 
 	percentageComplete := getPercentageComplete(tasksGetTaskResponse.Task.Status)
 
+	// If the task is marked complete but the percentage complete is not 1 then the task stopped unexpectedly
+	if tasksGetTaskResponse.Completed && percentageComplete != 1 {
+		return project_update_lib.JobStatus{}, fmt.Errorf("Task %s stopped unexpectedly. "+
+			"For more information go to http://localhost:10141/.tasks/task/%s", jobID, jobID)
```
Comment (tylercloke) [inline components/compliance-service/ingest/ingestic/ingestic.go]: This seems like it might not be the best info in some multi-node world but it's fine for now.

File: components/compliance-service/ingest/ingestic/ingestic.go (hunk 4 above)
Comment (lancewf) [reply components/compliance-service/ingest/ingestic/ingestic.go]: How do you create a function in the painless script?

File: components/compliance-service/ingest/ingestic/ingestic.go [hunk 6]
```diff
@@ -513,26 +546,37 @@ func (backend *ESClient) JobStatus(ctx context.Context, jobID string) (project_u
 	}, nil
 }
 
-func getPercentageComplete(status interface{}) float64 {
+func getPercentageComplete(status interface{}) (float64, bool) {
```
Comment (lancewf) [inline components/compliance-service/ingest/ingestic/ingestic.go]: This function is used for other types of async jobs. Therefore we need to add the create and deletes to the completed amounts.

Comment (lancewf) [reply components/compliance-service/ingest/ingestic/ingestic.go]: There is also the case that the "created", "deleted", and "updated" are missing from the status. In that case, we return a false to signal that the percentage should not be used, but it is not an error.

File: components/compliance-service/integration_test/reports_projects_update_test.go (hunk 3 above)
Comment (bcmdarroch) [reply components/compliance-service/integration_test/reports_projects_update_test.go]: Not sure if this should be a refactor or an additional test but it'd be great to see a test struct with all possible combinations of `attribute`, `operator`, and `value` (array or single value), to make sure we're not missing any edge cases. We could use two fully-populated `&relaxting.ESInSpecSummary` structs, one expected to match and one not to match, and iterate over the condition combinations on each test run.

Comment (lancewf) [reply components/compliance-service/integration_test/reports_projects_update_test.go]: I would like to cover all the cases. I don't fully understand what you are proposing, but it sounds interesting. 
The reason you don't see the `operator` in these tests is because it is completely ignored in the code.
